src/RequestHandler/outofprocess/forwardinghandler.cpp

```diff
@@ -536,103 +555,131 @@ REQUEST_NOTIFICATION_STATUS
-            // Honor the response status.
-            //
-        }
-    }
-    else
-    {
-        STACK_STRU(strDescription, 128);
+            STACK_STRU(strDescription, 128);
 
-        pResponse->SetStatus(502, "Bad Gateway", 3, hr);
+            pResponse->SetStatus(502, "Bad Gateway", 3, hr);
 
-        if (!(hr > HRESULT_FROM_WIN32(WINHTTP_ERROR_BASE) &&
-            hr <= HRESULT_FROM_WIN32(WINHTTP_ERROR_LAST)) ||
+            if (hr > HRESULT_FROM_WIN32(WINHTTP_ERROR_BASE) &&
+                hr <= HRESULT_FROM_WIN32(WINHTTP_ERROR_LAST))
+            {
 #pragma prefast (suppress : __WARNING_FUNCTION_NEEDS_REVIEW, "Function and parameters reviewed.")
-            FormatMessage(
-                FORMAT_MESSAGE_IGNORE_INSERTS | FORMAT_MESSAGE_FROM_HMODULE,
-                g_hWinHttpModule,
-                HRESULT_CODE(hr),
-                0,
-                strDescription.QueryStr(),
-                strDescription.QuerySizeCCH(),
-                NULL) == 0)
-        {
-            LoadString(g_hAspNetCoreModule,
-                IDS_SERVER_ERROR,
-                strDescription.QueryStr(),
-                strDescription.QuerySizeCCH());
-        }
+                FormatMessage(
+                    FORMAT_MESSAGE_IGNORE_INSERTS | FORMAT_MESSAGE_FROM_HMODULE,
+                    g_hWinHttpModule,
+                    HRESULT_CODE(hr),
+                    0,
+                    strDescription.QueryStr(),
+                    strDescription.QuerySizeCCH(),
+                    NULL);
+            }
+            else
+            {
+                LoadString(g_hAspNetCoreModule,
+                    IDS_SERVER_ERROR,
+                    strDescription.QueryStr(),
+                    strDescription.QuerySizeCCH());
+            }
+            (VOID)strDescription.SyncWithBuffer();
 
-        (VOID)strDescription.SyncWithBuffer();
-        if (strDescription.QueryCCH() != 0)
-        {
-            pResponse->SetErrorDescription(
-                strDescription.QueryStr(),
-                strDescription.QueryCCH(),
-                FALSE);
+            if (strDescription.QueryCCH() != 0)
+            {
+                pResponse->SetErrorDescription(
+                    strDescription.QueryStr(),
+                    strDescription.QueryCCH(),
+                    FALSE);
+            }
+
+            if (hr == HRESULT_FROM_WIN32(ERROR_WINHTTP_INVALID_SERVER_RESPONSE))
+            {
+                if (!m_fServerResetConn)
+                {
+                    RemoveRequest();
+                    pResponse->ResetConnection();
+                    m_fServerResetConn = TRUE;
+                }
+            }
         }
     }
 
-    //
-    // Finish the request on failure.
-    // Let IIS pipeline continue only after receiving handle close callback
-    // from WinHttp. This ensures no more callback from WinHttp
-    //
-    if (m_hRequest != NULL)
+    if (m_pWebSocket != NULL && !m_fWebSocketHandleInClose)
```

Comment:
Do you need to set m_pWebsocket to NULL here?

Reply:
no. We need call m_pWebSocket->Terminate() in the destructor to free the memory